proposals/0836-lean-core-jsc.md

```diff
@@ -0,0 +1,33 @@
+---
+title: Extract JSC (JavaScriptCore) from React Native Core
+author:
+- Kudo Chien <kudo@expo.dev>
+date: 2024-11-28
+---
+
+# RFC0836: Extract JSC (JavaScriptCore) from React Native Core
+
+## Summary
+
+This RFC proposes a Lean Core effort to extract JSC (JavaScriptCore) from React Native's core.
+
+## Proposed Changes and Timeline
+
+- Version 0.77: Remove the [`jsc-android` npm dependency](https://github.com/facebook/react-native/blob/2d337efc23c662143b3f39a6c994d80fec047054/packages/react-native/package.json#L132) from React Native and instead download the artifact from Maven Central. When using Hermes, downloading the unused npm dependency is unnecessary and removing it saves about 32 MB of uncompressed size.
+- Version 0.78 or 0.79:  Introduce a new `react-native-javascriptcore` npm package with a newer version of JavaScriptCore for Android. In this version, we will only provide the [Intl variant](https://github.com/react-native-community/jsc-android-buildscripts/blob/9c61fece4753902a2cd6d29dfa46b7b521f0c821/README.md#international-variant), as newer JavaScriptCore versions do not support disabling Intl. The existing JSCRuntime from the core should still function with this version but will display a deprecation warning during build time. A challenge will be managing the core JSCRuntime alongside the third-party JSCRuntime from `react-native-javascriptcore`; further study is required.
+- Version 0.80: Completely remove JavaScriptCore from the core:
+  - Eliminate `JSCRuntime` code.
+  - Deprecate Gradle's `hermesEnabled` and Podfile's `hermes_enabled`, making them no-ops.
+  - Update documentation at https://reactnative.dev/docs/hermes.
+
+## Motivation
+
+Since Hermes has become the dominant JavaScript engine in the React Native ecosystem, it is inefficient to include unused JavaScriptCore code and binaries in the core. By extracting JavaScriptCore to a separate, third-party maintained library, we can streamline the core and reduce unnecessary bloat.
+
+## Adoption strategy
+
+We will offer a drop-in replacement with the `react-native-javascriptcore` npm package while maintaining JSC support in the core for two additional React Native releases to ensure a smooth transition.
```

Comment:
If the boundary is jsi::Runtime, then I think that means there could be frequent updates needed against RN package.

The current structure has a nice benefit of keeping it up to date automatically, since it must build in line with JSI changes. But I think it means the project would need a long term maintainer to continue working. See https://github.com/microsoft/v8-jsi/commit/44384663695d4fd9e77464e86ce5581d2c4c8345

Depending on whether there is a natural maintainer, the other option would look like a harder deprecation, which would still allow externally maintained runtimes to develop.

Reply:
this is also the same problem for https://github.com/Kudo/react-native-v8. i can try to maintain the in my spare time (also love to have other maintainers from community)

maybe we need to include nightly testing from v8 and react-native-javascriptcore, so that we can catch jsi interface breaking changes and take action earlier.

Reply:
We'd be happy to help with maintenance of that RN <-> JSC integration at Callstack

Reply:
awesome! love the collaboration with someone from Callstack then.